File: plugins/dissector/dissector-essentials/ipv6.cpp
#include <nan.h>
#include <plugkit/dissector.h>
#include <plugkit/context.h>
#include <plugkit/token.h>
#include <plugkit/attribute.h>
#include <plugkit/variant.h>
#include <plugkit/layer.h>
#include <plugkit/payload.h>
#include <plugkit/reader.h>
#include <unordered_map>

using namespace plugkit;

namespace {

const std::unordered_map<uint16_t, std::pair<Token, Token>> protoTable = {
    {0x01,
     std::make_pair(Token_get("[icmp]"), Token_get("ipv6.protocol.icmp"))},
    {0x02,
     std::make_pair(Token_get("[igmp]"), Token_get("ipv6.protocol.igmp"))},
    {0x06, std::make_pair(Token_get("[tcp]"), Token_get("ipv6.protocol.tcp"))},
    {0x11, std::make_pair(Token_get("[udp]"), Token_get("ipv6.protocol.udp"))},
};

const auto ipv6Token = Token_get("ipv6");
const auto versionToken = Token_get("ipv6.version");
const auto tClassToken = Token_get("ipv6.tClass");
const auto fLevelToken = Token_get("ipv6.fLevel");
const auto pLenToken = Token_get("ipv6.pLen");
const auto nHeaderToken = Token_get("ipv6.nHeader");
const auto hLimitToken = Token_get("ipv6.hLimit");
const auto srcToken = Token_get(".src");
const auto dstToken = Token_get(".dst");
const auto hbyhToken = Token_get("ipv6.hbyh");
const auto protocolToken = Token_get("ipv6.protocol");
const auto ipv6AddrToken = Token_get("@ipv6:addr");
const auto flagsTypeToken = Token_get("@flags");
const auto enumToken = Token_get("@enum");

void analyze(Context *ctx, void *data, Layer *layer) {
  Reader reader;
  Reader_reset(&reader);
  reader.slice = Payload_slice(Layer_payload(layer));

  Layer *child = Layer_addLayer(layer, ipv6Token);
  Layer_addTag(child, ipv6Token);

  uint8_t header = Reader_readUint8(&reader);
  uint8_t header2 = Reader_readUint8(&reader);
  int version = header >> 4;
  int trafficClass = (header & 0b00001111 << 4) | ((header2 & 0b11110000) >> 4);
  int flowLevel = Reader_readUint16BE(&reader) | ((header2 & 0b00001111) << 16);

  Attr *ver = Layer_addAttr(child, versionToken);
  Attr_setUint32(ver, version);
  Attr_setRange(ver, Range{0, 1});

  Attr *tClass = Layer_addAttr(child, tClassToken);
  Attr_setUint32(tClass, trafficClass);
  Attr_setRange(tClass, Range{0, 2});

  Attr *fLevel = Layer_addAttr(child, fLevelToken);
  Attr_setUint32(fLevel, flowLevel);
  Attr_setRange(fLevel, Range{1, 4});

  Attr *pLen = Layer_addAttr(child, pLenToken);
  Attr_setUint32(pLen, Reader_readUint16BE(&reader));
  Attr_setRange(pLen, reader.lastRange);

  int nextHeader = Reader_readUint8(&reader);
  auto nextHeaderRange = reader.lastRange;

  Attr *nHeader = Layer_addAttr(child, nHeaderToken);
  Attr_setUint32(nHeader, nextHeader);
  Attr_setRange(nHeader, nextHeaderRange);

  Attr *hLimit = Layer_addAttr(child, hLimitToken);
  Attr_setUint32(hLimit, Reader_readUint8(&reader));
  Attr_setRange(hLimit, reader.lastRange);

  const auto &srcSlice = Reader_slice(&reader, 0, 16);
  Attr *src = Layer_addAttr(child, srcToken);
  Attr_setSlice(src, srcSlice);
  Attr_setType(src, ipv6AddrToken);
  Attr_setRange(src, reader.lastRange);

  const auto &dstSlice = Reader_slice(&reader, 0, 16);
  Attr *dst = Layer_addAttr(child, dstToken);
  Attr_setSlice(dst, dstSlice);
  Attr_setType(dst, ipv6AddrToken);
  Attr_setRange(dst, reader.lastRange);

  bool ext = true;
  while (ext) {
    int header = 0;
    switch (nextHeader) {
    case 0:
    case 60: // Hop-by-Hop Options, Destination Options
    {
      header = Reader_readUint8(&reader);
      size_t extLen = Reader_readUint8(&reader);
      size_t byteLen = (extLen + 1) * 8;
      Reader_slice(&reader, 0, byteLen);
      Token id = (nextHeader == 0) ? hbyhToken : dstToken;
    }

    break;
    // TODO:
    // case 43  # Routing
    // case 44  # Fragment
    // case 51  # Authentication Header
    // case 50  # Encapsulating Security Payload
    // case 135 # Mobility
    case 59: // No Next Header
    default:
      ext = false;
      continue;
    }

    nextHeader = header;
  }

  uint8_t protocolNumber = nextHeader;
  Attr *proto = Layer_addAttr(child, protocolToken);
  Attr_setUint32(proto, protocolNumber);
  Attr_setType(proto, enumToken);
  Attr_setRange(proto, reader.lastRange);
  const auto &it = protoTable.find(protocolNumber);
  if (it != protoTable.end()) {
    Attr *sub = Layer_addAttr(child, it->second.second);
    Attr_setBool(sub, true);
    Attr_setRange(sub, reader.lastRange);
    Layer_addTag(child, it->second.first);
  }

  Payload *chunk = Layer_addPayload(child);
  Payload_addSlice(chunk, Reader_sliceAll(&reader, 0));
}
}

void Init(v8::Local<v8::Object> exports) {
  Dissector *diss = Dissector_create(DISSECTOR_PACKET);
  Dissector_addLayerHint(diss, Token_get("[ipv6]"));
  Dissector_setAnalyzer(diss, analyze);
  exports->Set(Nan::New("dissector").ToLocalChecked(),
               Nan::New<v8::External>(diss));
}

NODE_MODULE(dissectorEssentials, Init);

File: package/ipv6/src/ipv6.rs
extern crate byteorder;
extern crate libc;

#[macro_use]
extern crate plugkit;

use std::io::{Cursor, Error, ErrorKind};
use byteorder::BigEndian;
use plugkit::reader::ByteReader;
use plugkit::layer::{Confidence, Layer};
use plugkit::context::Context;
use plugkit::worker::Worker;
use plugkit::variant::Value;
use plugkit::token::Token;
use plugkit::attr::ResultValue;

fn get_protocol(val: u32) -> Option<(Token, Token)> {
    match val {
        0x01 => Some((token!("[icmp]"), token!("ipv6.protocol.icmp"))),
        0x02 => Some((token!("[igmp]"), token!("ipv6.protocol.igmp"))),
        0x06 => Some((token!("[tcp]"), token!("ipv6.protocol.tcp"))),
        0x11 => Some((token!("[udp]"), token!("ipv6.protocol.udp"))),
        _ => None,
    }
}

struct IPv6Worker {}

impl Worker for IPv6Worker {
    fn analyze(&self, ctx: &mut Context, layer: &mut Layer) -> Result<(), Error> {
        let (slice, payload_range) = {
            let payload = layer
                .payloads()
                .next()
                .ok_or(Error::new(ErrorKind::Other, "no payload"))?;
            let slice = payload
                .slices()
                .next()
                .ok_or(Error::new(ErrorKind::Other, "no slice"))?;
            (slice, payload.range())
        };

        let child = layer.add_layer(ctx, token!("ipv6"));
        child.set_confidence(Confidence::Error);
        child.add_tag(token!("ipv6"));
        child.set_range(&payload_range);

        let mut rdr = Cursor::new(slice);
        let (header, _) = ByteReader::read_u8(&mut rdr)?;
        let (header2, _) = ByteReader::read_u8(&mut rdr)?;

        let version = header >> 4;
        let traffic_class = (header & 0b00001111 << 4) | ((header2 & 0b11110000) >> 4);

        {
            let attr = child.add_attr(ctx, token!("ipv6.version"));
            attr.set(&version);
            attr.set_range(&(0..1));
        }
        {
            let attr = child.add_attr(ctx, token!("ipv6.trafficClass"));
            attr.set(&traffic_class);
            attr.set_range(&(0..2));
        }

        let (flow_level, _) = ByteReader::read_u16::<BigEndian>(&mut rdr)?;
        {
            let attr = child.add_attr(ctx, token!("ipv6.flowLevel"));
            attr.set(&(flow_level | ((header2 & 0b00001111) << 16)));
            attr.set_range(&(1..4));
        }
        {
            let attr = child.add_attr(ctx, token!("ipv6.payloadLength"));
            attr.set_result(ByteReader::read_u16::<BigEndian>(&mut rdr))?;
        }

        let (mut next_header, mut next_range) = ByteReader::read_u8(&mut rdr)?;
        {
            let attr = child.add_attr(ctx, token!("ipv6.nextHeader"));
            attr.set(&next_header);
            attr.set_range(&next_range);
        }
        {
            let attr = child.add_attr(ctx, token!("ipv6.hopLimit"));
            attr.set_result(ByteReader::read_u8(&mut rdr))?;
        }

        {
            let attr = child.add_attr(ctx, token!("ipv6.src"));
            attr.set_typ(token!("@ipv6:addr"));
            attr.set_result(ByteReader::read_slice(&mut rdr, 16))?;
        }
        {
            let attr = child.add_attr(ctx, token!("ipv6.dst"));
            attr.set_typ(token!("@ipv6:addr"));
            attr.set_result(ByteReader::read_slice(&mut rdr, 16))?;
        }

        loop {
            match next_header {
                // Hop-by-Hop Options, Destination Options
                0 | 60 => {
                    let (h, r) = ByteReader::read_u8(&mut rdr)?;
                    next_header = h;
                    next_range = r;
                    let (ext_len, _) = ByteReader::read_u8(&mut rdr)?;
                    let byte_len = (ext_len + 1) * 8;
                    ByteReader::read_slice(&mut rdr, byte_len as usize)?;
                },
                // TODO:
                // case 43  # Routing
                // case 44  # Fragment
                // case 51  # Authentication Header
                // case 50  # Encapsulating Security Payload
                // case 135 # Mobility
                // No Next Header
                59 => break,
                _ => break
            }
        }

        let protocol = next_header;
        let range = next_range;
        {
            let attr = child.add_attr(ctx, token!("ipv6.protocol"));
            attr.set_typ(token!("@enum"));
            attr.set(&protocol);
            attr.set_range(&range);
        }
        if let Some(item) = get_protocol(protocol) {
            let (tag, id) = item;
            child.add_tag(tag);
            let attr = child.add_attr(ctx, id);
            attr.set(&true);
            attr.set_typ(token!("@novalue"));
            attr.set_range(&range);
        }

        {
            let (data, range) = ByteReader::read_slice_to_end(&mut rdr)?;
            let payload = child.add_payload(ctx);
            let offset = payload_range.start;
            payload.add_slice(data);
            payload.set_range(&(range.start + offset..range.end + offset));
        }

        child.set_confidence(Confidence::Exact);
        Ok(())
    }
}

plugkit_module!({});
plugkit_api_layer_hints!(token!("[ipv6]"));
plugkit_api_worker!(IPv6Worker, IPv6Worker {});
